To run this, press "*Runtime*" and press "*Run all*" on a **free** Tesla T4 Google Colab instance!
<div class="align-center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a></a> Join Discord if you need help + ⭐ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐
</div>

To install Unsloth on your own computer, follow the installation instructions on our Github page [here](https://docs.unsloth.ai/get-started/installing-+-updating).

You will learn how to do [data prep](#Data), how to [train](#Train), how to [run the model](#Inference), & [how to save it](#Save)


### News

**Read our [Gemma 3 blog](https://unsloth.ai/blog/gemma3) for what's new in Unsloth and our [Reasoning blog](https://unsloth.ai/blog/r1-reasoning) on how to train reasoning models.**

Visit our docs for all our [model uploads](https://docs.unsloth.ai/get-started/all-our-models) and [notebooks](https://docs.unsloth.ai/get-started/unsloth-notebooks).


### Installation

In [2]:
%%capture
import os
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth
else:
    # Do this only in Colab notebooks! Otherwise use pip install unsloth
    !pip install --no-deps bitsandbytes accelerate xformers==0.0.29.post3 peft trl==0.15.2 triton cut_cross_entropy unsloth_zoo
    !pip install sentencepiece protobuf datasets huggingface_hub hf_transfer
    !pip install --no-deps unsloth
!pip install snac


In [1]:
from huggingface_hub import notebook_login

notebook_login()

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


### Unsloth

`FastModel` supports loading nearly any model now! This includes Vision and Text models!

Thank you to [Etherl](https://huggingface.co/Etherll) for creating this notebook!

In [ ]:
from unsloth import FastLanguageModel
import torch
dtype = None # None for auto detection. Float16 for Tesla T4, V100, Bfloat16 for Ampere+
load_in_4bit = True # Use 4bit quantization to reduce memory usage. Can be False.
# 4bit pre quantized models we support for 4x faster downloading + no OOMs.
fourbit_models = [
    "unsloth/Meta-Llama-3.1-8B-bnb-4bit",      # Llama-3.1 15 trillion tokens model 2x faster!
    "unsloth/Meta-Llama-3.1-8B-Instruct-bnb-4bit",
    "unsloth/Meta-Llama-3.1-70B-bnb-4bit",
    "unsloth/Meta-Llama-3.1-405B-bnb-4bit",    # We also uploaded 4bit for 405b!
    "unsloth/Mistral-Nemo-Base-2407-bnb-4bit", # New Mistral 12b 2x faster!
    "unsloth/Mistral-Nemo-Instruct-2407-bnb-4bit",
    "unsloth/mistral-7b-v0.3-bnb-4bit",        # Mistral v3 2x faster!
    "unsloth/mistral-7b-instruct-v0.3-bnb-4bit",
    "unsloth/Phi-3.5-mini-instruct",           # Phi-3.5 2x faster!
    "unsloth/Phi-3-medium-4k-instruct",
    "unsloth/gemma-2-9b-bnb-4bit",
    "unsloth/gemma-2-27b-bnb-4bit",            # Gemma 2x faster!
] # More models at https://huggingface.co/unsloth

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "unsloth/orpheus-3b-0.1-ft-unsloth-bnb-4bit",
    max_seq_length= 2048, # Choose any for long context!
    dtype = dtype,
    load_in_4bit = load_in_4bit,
    #token = "hf_...", # use one if using gated models like meta-llama/Llama-2-7b-hf
)

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth Zoo will now patch everything to make training faster!
==((====))==  Unsloth 2025.3.19: Fast Llama patching. Transformers: 4.50.3.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 7.5. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.29.post3. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


In [ ]:
# --- 1. 导入必要的库 ---
from unsloth import FastLanguageModel
import torch

# --- 2. 设置模型加载参数 ---
# 您可以根据您的硬件和需求调整这些参数
dtype = None # 设置为 None 让库自动检测，或者 torch.float16, torch.bfloat16
load_in_4bit = True # 如果您的显存有限或者想加速加载，可以设为 True；否则设为 False

print("正在加载模型和分词器...")
print(f"模型名称: unsloth/orpheus-3b-0.1-ft-unsloth-bnb-4bit")
print(f"使用4bit加载: {load_in_4bit}")
print("这可能需要一些时间，取决于您的网络和硬件...")

# --- 3. 加载模型和分词器 ---
# 注意：加载模型会消耗较多内存/显存
try:
    model, tokenizer = FastLanguageModel.from_pretrained(
        model_name = "unsloth/orpheus-3b-0.1-ft-unsloth-bnb-4bit",
        max_seq_length= 2048, # 与您之前代码一致
        dtype = dtype,
        load_in_4bit = load_in_4bit,
        # 如果需要 Hugging Face token (例如 Llama 2)，取消下面一行的注释
        # token = "hf_...",
    )
    print("模型和分词器加载成功！")
except Exception as e:
    print(f"加载过程中出现错误: {e}")
    print("请检查模型名称、网络连接、Hugging Face token（如果需要）以及您的硬件资源。")
    # 如果加载失败，后续代码无法执行，直接退出或处理错误
    raise e # 重新抛出异常，以便在Jupyter中看到完整的错误栈

# --- 4. 定义测试句子 ---
test_sentence = "你好，世界！这是一个有趣的测试。"
# 您可以换成任何想测试的句子

# --- 5. 使用分词器进行分词 ---

# 获取 Token ID 列表
# add_special_tokens=False 表示不添加特殊的起始/结束符，只看句子本身的分词结果
# 如果想看添加特殊符号后的结果，可以设为 True
input_ids = tokenizer.encode(test_sentence, add_special_tokens=False)

# 将 Token ID 转换回 Token 字符串列表，方便查看
tokens = tokenizer.convert_ids_to_tokens(input_ids)

# 将 Token ID 解码回原始字符串，用于验证
decoded_sentence = tokenizer.decode(input_ids, skip_special_tokens=True)

# --- 6. 打印结果 ---
print("\n--- 分词测试结果 ---")
print(f"原始句子: {test_sentence}")
print(f"Token 字符串: {tokens}")
print(f"对应的 Token ID: {input_ids}")
print(f"解码后的句子: {decoded_sentence}")

# 额外提示：查看词汇表大小
vocab_size = tokenizer.vocab_size
print(f"\n分词器的词汇表大小: {vocab_size}")

# 额外提示：测试带特殊符号的情况
input_ids_with_special = tokenizer.encode(test_sentence, add_special_tokens=True)
tokens_with_special = tokenizer.convert_ids_to_tokens(input_ids_with_special)
print("\n--- 带特殊符号的分词结果 ---")
print(f"Token 字符串 (带特殊符号): {tokens_with_special}")
print(f"对应的 Token ID (带特殊符号): {input_ids_with_special}")

We now add LoRA adapters so we only need to update 1 to 10% of all parameters!

In [ ]:
model = FastLanguageModel.get_peft_model(
    model,
    r = 64, # Choose any number > 0 ! Suggested 8, 16, 32, 64, 128
    target_modules = ["q_proj", "k_proj", "v_proj", "o_proj",
                      "gate_proj", "up_proj", "down_proj",],
    lora_alpha = 64,
    lora_dropout = 0, # Supports any, but = 0 is optimized
    bias = "none",    # Supports any, but = "none" is optimized
    # [NEW] "unsloth" uses 30% less VRAM, fits 2x larger batch sizes!
    use_gradient_checkpointing = "unsloth", # True or "unsloth" for very long context
    random_state = 3407,
    use_rslora = False,  # We support rank stabilized LoRA
    loftq_config = None, # And LoftQ
)

Unsloth 2025.3.19 patched 28 layers with 28 QKV layers, 28 O layers and 28 MLP layers.


<a name="Data"></a>
### Data Prep  

We will use the `MrDragonFox/Elise`, which is designed for training TTS models. Ensure that your dataset follows the required format: **text, audio** for single-speaker models or **source, text, audio** for multi-speaker models. You can modify this section to accommodate your own dataset, but maintaining the correct structure is essential for optimal training.

In [3]:
from datasets import load_dataset
dataset = load_dataset("hirubyyyy/rubys_all-speakers", split = "train")

README.md:   0%|          | 0.00/326 [00:00<?, ?B/s]

Resolving data files:   0%|          | 0/136 [00:00<?, ?it/s]

Resolving data files:   0%|          | 0/136 [00:00<?, ?it/s]

train-00000-of-00136.parquet:   0%|          | 0.00/490M [00:00<?, ?B/s]

train-00001-of-00136.parquet:   0%|          | 0.00/845M [00:00<?, ?B/s]

train-00002-of-00136.parquet:   0%|          | 0.00/922M [00:00<?, ?B/s]

train-00003-of-00136.parquet:   0%|          | 0.00/910M [00:00<?, ?B/s]

train-00004-of-00136.parquet:   0%|          | 0.00/812M [00:00<?, ?B/s]

train-00005-of-00136.parquet:   0%|          | 0.00/1.05G [00:00<?, ?B/s]

train-00006-of-00136.parquet:   0%|          | 0.00/946M [00:00<?, ?B/s]

train-00007-of-00136.parquet:   0%|          | 0.00/1.11G [00:00<?, ?B/s]

train-00008-of-00136.parquet:   0%|          | 0.00/865M [00:00<?, ?B/s]

train-00009-of-00136.parquet:   0%|          | 0.00/471M [00:00<?, ?B/s]

train-00010-of-00136.parquet:   0%|          | 0.00/708M [00:00<?, ?B/s]

train-00011-of-00136.parquet:   0%|          | 0.00/471M [00:00<?, ?B/s]

train-00012-of-00136.parquet:   0%|          | 0.00/677M [00:00<?, ?B/s]

train-00013-of-00136.parquet:   0%|          | 0.00/1.09G [00:00<?, ?B/s]

train-00014-of-00136.parquet:   0%|          | 0.00/796M [00:00<?, ?B/s]

train-00015-of-00136.parquet:   0%|          | 0.00/424M [00:00<?, ?B/s]

train-00016-of-00136.parquet:   0%|          | 0.00/686M [00:00<?, ?B/s]

train-00017-of-00136.parquet:   0%|          | 0.00/773M [00:00<?, ?B/s]

train-00018-of-00136.parquet:   0%|          | 0.00/808M [00:00<?, ?B/s]

train-00019-of-00136.parquet:   0%|          | 0.00/463M [00:00<?, ?B/s]

train-00020-of-00136.parquet:   0%|          | 0.00/1.19G [00:00<?, ?B/s]

train-00021-of-00136.parquet:   0%|          | 0.00/429M [00:00<?, ?B/s]

train-00022-of-00136.parquet:   0%|          | 0.00/704M [00:00<?, ?B/s]

train-00023-of-00136.parquet:   0%|          | 0.00/669M [00:00<?, ?B/s]

train-00024-of-00136.parquet:   0%|          | 0.00/458M [00:00<?, ?B/s]

train-00025-of-00136.parquet:   0%|          | 0.00/764M [00:00<?, ?B/s]

train-00026-of-00136.parquet:   0%|          | 0.00/1.04G [00:00<?, ?B/s]

train-00027-of-00136.parquet:   0%|          | 0.00/446M [00:00<?, ?B/s]

train-00028-of-00136.parquet:   0%|          | 0.00/794M [00:00<?, ?B/s]

train-00029-of-00136.parquet:   0%|          | 0.00/1.06G [00:00<?, ?B/s]

train-00030-of-00136.parquet:   0%|          | 0.00/512M [00:00<?, ?B/s]

train-00031-of-00136.parquet:   0%|          | 0.00/1.35G [00:00<?, ?B/s]

train-00032-of-00136.parquet:   0%|          | 0.00/828M [00:00<?, ?B/s]

train-00033-of-00136.parquet:   0%|          | 0.00/921M [00:00<?, ?B/s]

train-00034-of-00136.parquet:   0%|          | 0.00/903M [00:00<?, ?B/s]

train-00035-of-00136.parquet:   0%|          | 0.00/482M [00:00<?, ?B/s]

train-00036-of-00136.parquet:   0%|          | 0.00/761M [00:00<?, ?B/s]

train-00037-of-00136.parquet:   0%|          | 0.00/880M [00:00<?, ?B/s]

train-00038-of-00136.parquet:   0%|          | 0.00/586M [00:00<?, ?B/s]

train-00039-of-00136.parquet:   0%|          | 0.00/570M [00:00<?, ?B/s]

train-00040-of-00136.parquet:   0%|          | 0.00/803M [00:00<?, ?B/s]

train-00041-of-00136.parquet:   0%|          | 0.00/1.08G [00:00<?, ?B/s]

train-00042-of-00136.parquet:   0%|          | 0.00/459M [00:00<?, ?B/s]

train-00043-of-00136.parquet:   0%|          | 0.00/697M [00:00<?, ?B/s]

train-00044-of-00136.parquet:   0%|          | 0.00/944M [00:00<?, ?B/s]

train-00045-of-00136.parquet:   0%|          | 0.00/561M [00:00<?, ?B/s]

train-00046-of-00136.parquet:   0%|          | 0.00/781M [00:00<?, ?B/s]

train-00047-of-00136.parquet:   0%|          | 0.00/947M [00:00<?, ?B/s]

train-00048-of-00136.parquet:   0%|          | 0.00/1.29G [00:00<?, ?B/s]

train-00049-of-00136.parquet:   0%|          | 0.00/490M [00:00<?, ?B/s]

train-00050-of-00136.parquet:   0%|          | 0.00/977M [00:00<?, ?B/s]

train-00051-of-00136.parquet:   0%|          | 0.00/658M [00:00<?, ?B/s]

train-00052-of-00136.parquet:   0%|          | 0.00/746M [00:00<?, ?B/s]

train-00053-of-00136.parquet:   0%|          | 0.00/532M [00:00<?, ?B/s]

train-00054-of-00136.parquet:   0%|          | 0.00/738M [00:00<?, ?B/s]

train-00055-of-00136.parquet:   0%|          | 0.00/833M [00:00<?, ?B/s]

train-00056-of-00136.parquet:   0%|          | 0.00/1.19G [00:00<?, ?B/s]

train-00057-of-00136.parquet:   0%|          | 0.00/1.40G [00:00<?, ?B/s]

train-00058-of-00136.parquet:   0%|          | 0.00/514M [00:00<?, ?B/s]

train-00059-of-00136.parquet:   0%|          | 0.00/481M [00:00<?, ?B/s]

train-00060-of-00136.parquet:   0%|          | 0.00/624M [00:00<?, ?B/s]

train-00061-of-00136.parquet:   0%|          | 0.00/857M [00:00<?, ?B/s]

train-00062-of-00136.parquet:   0%|          | 0.00/445M [00:00<?, ?B/s]

train-00063-of-00136.parquet:   0%|          | 0.00/775M [00:00<?, ?B/s]

train-00064-of-00136.parquet:   0%|          | 0.00/850M [00:00<?, ?B/s]

train-00065-of-00136.parquet:   0%|          | 0.00/749M [00:00<?, ?B/s]

train-00066-of-00136.parquet:   0%|          | 0.00/458M [00:00<?, ?B/s]

train-00067-of-00136.parquet:   0%|          | 0.00/1.01G [00:00<?, ?B/s]

train-00068-of-00136.parquet:   0%|          | 0.00/806M [00:00<?, ?B/s]

train-00069-of-00136.parquet:   0%|          | 0.00/640M [00:00<?, ?B/s]

train-00070-of-00136.parquet:   0%|          | 0.00/860M [00:00<?, ?B/s]

train-00071-of-00136.parquet:   0%|          | 0.00/447M [00:00<?, ?B/s]

train-00072-of-00136.parquet:   0%|          | 0.00/816M [00:00<?, ?B/s]

train-00073-of-00136.parquet:   0%|          | 0.00/630M [00:00<?, ?B/s]

train-00074-of-00136.parquet:   0%|          | 0.00/906M [00:00<?, ?B/s]

train-00075-of-00136.parquet:   0%|          | 0.00/479M [00:00<?, ?B/s]

train-00076-of-00136.parquet:   0%|          | 0.00/826M [00:00<?, ?B/s]

train-00077-of-00136.parquet:   0%|          | 0.00/1.10G [00:00<?, ?B/s]

train-00078-of-00136.parquet:   0%|          | 0.00/815M [00:00<?, ?B/s]

train-00079-of-00136.parquet:   0%|          | 0.00/513M [00:00<?, ?B/s]

train-00080-of-00136.parquet:   0%|          | 0.00/980M [00:00<?, ?B/s]

train-00081-of-00136.parquet:   0%|          | 0.00/426M [00:00<?, ?B/s]

train-00082-of-00136.parquet:   0%|          | 0.00/861M [00:00<?, ?B/s]

train-00083-of-00136.parquet:   0%|          | 0.00/521M [00:00<?, ?B/s]

train-00084-of-00136.parquet:   0%|          | 0.00/919M [00:00<?, ?B/s]

train-00085-of-00136.parquet:   0%|          | 0.00/1.29G [00:00<?, ?B/s]

train-00086-of-00136.parquet:   0%|          | 0.00/823M [00:00<?, ?B/s]

train-00087-of-00136.parquet:   0%|          | 0.00/696M [00:00<?, ?B/s]

train-00088-of-00136.parquet:   0%|          | 0.00/668M [00:00<?, ?B/s]

train-00089-of-00136.parquet:   0%|          | 0.00/747M [00:00<?, ?B/s]

train-00090-of-00136.parquet:   0%|          | 0.00/1.22G [00:00<?, ?B/s]

train-00091-of-00136.parquet:   0%|          | 0.00/597M [00:00<?, ?B/s]

train-00092-of-00136.parquet:   0%|          | 0.00/719M [00:00<?, ?B/s]

train-00093-of-00136.parquet:   0%|          | 0.00/811M [00:00<?, ?B/s]

train-00094-of-00136.parquet:   0%|          | 0.00/931M [00:00<?, ?B/s]

train-00095-of-00136.parquet:   0%|          | 0.00/497M [00:00<?, ?B/s]

train-00096-of-00136.parquet:   0%|          | 0.00/508M [00:00<?, ?B/s]

train-00097-of-00136.parquet:   0%|          | 0.00/1.09G [00:00<?, ?B/s]

/usr/local/lib/python3.11/dist-packages/huggingface_hub/file_download.py:752: UserWarning: Not enough free disk space to download the file. The expected file size is: 711.45 MB. The target location /root/.cache/huggingface/hub/datasets--hirubyyyy--rubys_all-speakers/blobs only has 649.78 MB free disk space.
  warnings.warn(


train-00098-of-00136.parquet:   0%|          | 0.00/711M [00:00<?, ?B/s]

/usr/local/lib/python3.11/dist-packages/huggingface_hub/file_download.py:752: UserWarning: Not enough free disk space to download the file. The expected file size is: 954.55 MB. The target location /root/.cache/huggingface/hub/datasets--hirubyyyy--rubys_all-speakers/blobs only has 0.00 MB free disk space.
  warnings.warn(


train-00099-of-00136.parquet:   0%|          | 0.00/955M [00:00<?, ?B/s]

OSError: [Errno 28] No space left on device

In [ ]:
#@title Tokenisation Function

import locale
import torchaudio.transforms as T
import os
import torch
from snac import SNAC
locale.getpreferredencoding = lambda: "UTF-8"
ds_sample_rate = dataset[0]["audio"]["sampling_rate"]

snac_model = SNAC.from_pretrained("hubertsiuzdak/snac_24khz")
snac_model = snac_model.to("cuda")
def tokenise_audio(waveform):
  waveform = torch.from_numpy(waveform).unsqueeze(0)
  waveform = waveform.to(dtype=torch.float32)
  resample_transform = T.Resample(orig_freq=ds_sample_rate, new_freq=24000)
  waveform = resample_transform(waveform)

  waveform = waveform.unsqueeze(0).to("cuda")

  #generate the codes from snac
  with torch.inference_mode():
    codes = snac_model.encode(waveform)

  all_codes = []
  for i in range(codes[0].shape[1]):
    all_codes.append(codes[0][0][i].item()+128266)
    all_codes.append(codes[1][0][2*i].item()+128266+4096)
    all_codes.append(codes[2][0][4*i].item()+128266+(2*4096))
    all_codes.append(codes[2][0][(4*i)+1].item()+128266+(3*4096))
    all_codes.append(codes[1][0][(2*i)+1].item()+128266+(4*4096))
    all_codes.append(codes[2][0][(4*i)+2].item()+128266+(5*4096))
    all_codes.append(codes[2][0][(4*i)+3].item()+128266+(6*4096))


  return all_codes

def add_codes(example):
    # Always initialize codes_list to None
    codes_list = None

    try:
        answer_audio = example.get("audio")
        # If there's a valid audio array, tokenise it
        if answer_audio and "array" in answer_audio:
            audio_array = answer_audio["array"]
            codes_list = tokenise_audio(audio_array)
    except Exception as e:
        print(f"Skipping row due to error: {e}")
        # Keep codes_list as None if we fail
    example["codes_list"] = codes_list

    return example

dataset = dataset.map(add_codes, remove_columns=["audio"])

tokeniser_length = 128256
start_of_text = 128000
end_of_text = 128009

start_of_speech = tokeniser_length + 1
end_of_speech = tokeniser_length + 2

start_of_human = tokeniser_length + 3
end_of_human = tokeniser_length + 4

start_of_ai = tokeniser_length + 5
end_of_ai =  tokeniser_length + 6
pad_token = tokeniser_length + 7

audio_tokens_start = tokeniser_length + 10

dataset = dataset.filter(lambda x: x["codes_list"] is not None)
dataset = dataset.filter(lambda x: len(x["codes_list"]) > 0)

def remove_duplicate_frames(example):
    vals = example["codes_list"]
    if len(vals) % 7 != 0:
        raise ValueError("Input list length must be divisible by 7")

    result = vals[:7]

    removed_frames = 0

    for i in range(7, len(vals), 7):
        current_first = vals[i]
        previous_first = result[-7]

        if current_first != previous_first:
            result.extend(vals[i:i+7])
        else:
            removed_frames += 1

    example["codes_list"] = result

    return example

dataset = dataset.map(remove_duplicate_frames)

tok_info = '''*** HERE you can modify the text prompt
If you are training a multi-speaker model (e.g., canopylabs/orpheus-3b-0.1-ft),
ensure that the dataset includes a "source" field and format the input accordingly:
- Single-speaker: f"{example['text']}"
- Multi-speaker: f"{example['source']}: {example['text']}"
'''
print(tok_info)

def create_input_ids(example):
    # Determine whether to include the source field
    text_prompt = f"{example['source']}: {example['text']}" if "source" in example else example["text"]

    text_ids = tokenizer.encode(text_prompt, add_special_tokens=True)
    text_ids.append(end_of_text)

    example["text_tokens"] = text_ids
    input_ids = (
        [start_of_human]
        + example["text_tokens"]
        + [end_of_human]
        + [start_of_ai]
        + [start_of_speech]
        + example["codes_list"]
        + [end_of_speech]
        + [end_of_ai]
    )
    example["input_ids"] = input_ids
    example["labels"] = input_ids
    example["attention_mask"] = [1] * len(input_ids)

    return example


dataset = dataset.map(create_input_ids, remove_columns=["text", "codes_list"])
columns_to_keep = ["input_ids", "labels", "attention_mask"]
columns_to_remove = [col for col in dataset.column_names if col not in columns_to_keep]

dataset = dataset.remove_columns(columns_to_remove)

Parameter 'function'=<function add_codes at 0x7cf71a76b060> of the transform datasets.arrow_dataset.Dataset._map_single couldn't be hashed properly, a random hash was used instead. Make sure your transforms and parameters are serializable with pickle or dill for the dataset fingerprinting and caching to work. If you reuse this transform, the caching mechanism will consider it to be different from the previous calls and recompute everything. This warning is only showed once. Subsequent hashing failures won't be showed.


Map:   0%|          | 0/3588 [00:00<?, ? examples/s]

Skipping row due to error: cannot reshape tensor of 0 elements into shape [-1, 0] because the unspecified dimension size -1 can be any value and is ambiguous
Skipping row due to error: cannot reshape tensor of 0 elements into shape [-1, 0] because the unspecified dimension size -1 can be any value and is ambiguous
Skipping row due to error: cannot reshape tensor of 0 elements into shape [-1, 0] because the unspecified dimension size -1 can be any value and is ambiguous
Skipping row due to error: cannot reshape tensor of 0 elements into shape [-1, 0] because the unspecified dimension size -1 can be any value and is ambiguous
Skipping row due to error: cannot reshape tensor of 0 elements into shape [-1, 0] because the unspecified dimension size -1 can be any value and is ambiguous
Skipping row due to error: cannot reshape tensor of 0 elements into shape [-1, 0] because the unspecified dimension size -1 can be any value and is ambiguous
Skipping row due to error: cannot reshape tensor of 

Filter:   0%|          | 0/3588 [00:00<?, ? examples/s]

Filter:   0%|          | 0/3580 [00:00<?, ? examples/s]

Map:   0%|          | 0/3580 [00:00<?, ? examples/s]

*** HERE you can modify the text prompt
If you are training a multi-speaker model (e.g., canopylabs/orpheus-3b-0.1-ft),
ensure that the dataset includes a "source" field and format the input accordingly:
- Single-speaker: f"{example['text']}"
- Multi-speaker: f"{example['source']}: {example['text']}"



Map:   0%|          | 0/3580 [00:00<?, ? examples/s]

<a name="Train"></a>
### Train the model
Now let's use Huggingface  `Trainer`! More docs here: [Transformers docs](https://huggingface.co/docs/transformers/main_classes/trainer). We do 60 steps to speed things up, but you can set `num_train_epochs=1` for a full run, and turn off `max_steps=None`.

**Note:** Using a per_device_train_batch_size >1 may lead to errors if multi-GPU setup to avoid issues, ensure CUDA_VISIBLE_DEVICES is set to a single GPU (e.g., CUDA_VISIBLE_DEVICES=0).

In [ ]:
from transformers import TrainingArguments,Trainer,DataCollatorForSeq2Seq
from unsloth import is_bfloat16_supported

trainer = Trainer(
    model = model,
    train_dataset = dataset,
    args = TrainingArguments(
        per_device_train_batch_size = 1,
        gradient_accumulation_steps = 4,
        warmup_steps = 5,
        # num_train_epochs = 1, # Set this for 1 full training run.
        max_steps = 60,
        learning_rate = 2e-4,
        fp16 = not is_bfloat16_supported(),
        bf16 = is_bfloat16_supported(),
        logging_steps = 1,
        optim = "adamw_8bit",
        weight_decay = 0.01,
        lr_scheduler_type = "linear",
        seed = 3407,
        output_dir = "outputs",
        report_to = "none", # Use this for WandB etc
    ),
)

In [ ]:
# @title Show current memory stats
gpu_stats = torch.cuda.get_device_properties(0)
start_gpu_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
max_memory = round(gpu_stats.total_memory / 1024 / 1024 / 1024, 3)
print(f"GPU = {gpu_stats.name}. Max memory = {max_memory} GB.")
print(f"{start_gpu_memory} GB of memory reserved.")

GPU = Tesla T4. Max memory = 14.741 GB.
2.994 GB of memory reserved.


In [ ]:
trainer_stats = trainer.train()

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 3,580 | Num Epochs = 1 | Total steps = 60
O^O/ \_/ \    Batch size per device = 1 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (1 x 4 x 1) = 4
 "-____-"     Trainable parameters = 97,255,424/3,000,000,000 (3.24% trained)


Unsloth: Will smartly offload gradients to save VRAM!


In [ ]:
# @title Show final memory and time stats
used_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
used_memory_for_lora = round(used_memory - start_gpu_memory, 3)
used_percentage = round(used_memory / max_memory * 100, 3)
lora_percentage = round(used_memory_for_lora / max_memory * 100, 3)
print(f"{trainer_stats.metrics['train_runtime']} seconds used for training.")
print(
    f"{round(trainer_stats.metrics['train_runtime']/60, 2)} minutes used for training."
)
print(f"Peak reserved memory = {used_memory} GB.")
print(f"Peak reserved memory for training = {used_memory_for_lora} GB.")
print(f"Peak reserved memory % of max memory = {used_percentage} %.")
print(f"Peak reserved memory for training % of max memory = {lora_percentage} %.")

173.6206 seconds used for training.
2.89 minutes used for training.
Peak reserved memory = 4.178 GB.
Peak reserved memory for training = 1.184 GB.
Peak reserved memory % of max memory = 28.343 %.
Peak reserved memory for training % of max memory = 8.032 %.


<a name="Inference"></a>
### Inference
Let's run the model! You can change the prompts

**[NEW] Try 2x faster inference in a free Colab for Llama-3.1 8b Instruct [here](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Unsloth_Studio.ipynb)**

In [ ]:
prompts = [
    "<whisper>I like you....<normal>but this is a secret",
]

chosen_voice = None # None for single-speaker

In [ ]:
#@title Run Inference


FastLanguageModel.for_inference(model) # Enable native 2x faster inference

# Moving snac_model cuda to cpu
snac_model.to("cpu")

prompts_ = [(f"{chosen_voice}: " + p) if chosen_voice else p for p in prompts]

all_input_ids = []

for prompt in prompts_:
  input_ids = tokenizer(prompt, return_tensors="pt").input_ids
  all_input_ids.append(input_ids)

start_token = torch.tensor([[ 128259]], dtype=torch.int64) # Start of human
end_tokens = torch.tensor([[128009, 128260]], dtype=torch.int64) # End of text, End of human

all_modified_input_ids = []
for input_ids in all_input_ids:
  modified_input_ids = torch.cat([start_token, input_ids, end_tokens], dim=1) # SOH SOT Text EOT EOH
  all_modified_input_ids.append(modified_input_ids)

all_padded_tensors = []
all_attention_masks = []
max_length = max([modified_input_ids.shape[1] for modified_input_ids in all_modified_input_ids])
for modified_input_ids in all_modified_input_ids:
  padding = max_length - modified_input_ids.shape[1]
  padded_tensor = torch.cat([torch.full((1, padding), 128263, dtype=torch.int64), modified_input_ids], dim=1)
  attention_mask = torch.cat([torch.zeros((1, padding), dtype=torch.int64), torch.ones((1, modified_input_ids.shape[1]), dtype=torch.int64)], dim=1)
  all_padded_tensors.append(padded_tensor)
  all_attention_masks.append(attention_mask)

all_padded_tensors = torch.cat(all_padded_tensors, dim=0)
all_attention_masks = torch.cat(all_attention_masks, dim=0)

input_ids = all_padded_tensors.to("cuda")
attention_mask = all_attention_masks.to("cuda")
generated_ids = model.generate(
      input_ids=input_ids,
      attention_mask=attention_mask,
      max_new_tokens=1200,
      do_sample=True,
      temperature=0.6,
      top_p=0.95,
      repetition_penalty=1.1,
      num_return_sequences=1,
      eos_token_id=128258,
     use_cache = True
  )
token_to_find = 128257
token_to_remove = 128258

token_indices = (generated_ids == token_to_find).nonzero(as_tuple=True)

if len(token_indices[1]) > 0:
    last_occurrence_idx = token_indices[1][-1].item()
    cropped_tensor = generated_ids[:, last_occurrence_idx+1:]
else:
    cropped_tensor = generated_ids

mask = cropped_tensor != token_to_remove

processed_rows = []

for row in cropped_tensor:
    masked_row = row[row != token_to_remove]
    processed_rows.append(masked_row)

code_lists = []

for row in processed_rows:
    row_length = row.size(0)
    new_length = (row_length // 7) * 7
    trimmed_row = row[:new_length]
    trimmed_row = [t - 128266 for t in trimmed_row]
    code_lists.append(trimmed_row)


def redistribute_codes(code_list):
  layer_1 = []
  layer_2 = []
  layer_3 = []
  for i in range((len(code_list)+1)//7):
    layer_1.append(code_list[7*i])
    layer_2.append(code_list[7*i+1]-4096)
    layer_3.append(code_list[7*i+2]-(2*4096))
    layer_3.append(code_list[7*i+3]-(3*4096))
    layer_2.append(code_list[7*i+4]-(4*4096))
    layer_3.append(code_list[7*i+5]-(5*4096))
    layer_3.append(code_list[7*i+6]-(6*4096))
  codes = [torch.tensor(layer_1).unsqueeze(0),
         torch.tensor(layer_2).unsqueeze(0),
         torch.tensor(layer_3).unsqueeze(0)]

  # codes = [c.to("cuda") for c in codes]
  audio_hat = snac_model.decode(codes)
  return audio_hat

my_samples = []
for code_list in code_lists:
  samples = redistribute_codes(code_list)
  my_samples.append(samples)
from IPython.display import display, Audio
if len(prompts) != len(my_samples):
  raise Exception("Number of prompts and samples do not match")
else:
  for i in range(len(my_samples)):
    print(prompts[i])
    samples = my_samples[i]
    display(Audio(samples.detach().squeeze().to("cpu").numpy(), rate=24000))
# Clean up to save RAM
del my_samples,samples

<whisper>I like you....<normal>but this is a secret


<a name="Save"></a>
### Saving, loading finetuned models
To save the final model as LoRA adapters, either use Huggingface's `push_to_hub` for an online save or `save_pretrained` for a local save.

**[NOTE]** This ONLY saves the LoRA adapters, and not the full model. To save to 16bit or GGUF, scroll down!

In [ ]:
model.save_pretrained("lora_model")  # Local saving
tokenizer.save_pretrained("lora_model")
# model.push_to_hub("your_name/lora_model", token = "...") # Online saving
# tokenizer.push_to_hub("your_name/lora_model", token = "...") # Online saving

('lora_model/tokenizer_config.json',
 'lora_model/special_tokens_map.json',
 'lora_model/tokenizer.json')

You can also use Hugging Face's `AutoModelForPeftCausalLM`. Only use this if you do not have `unsloth` installed. It can be hopelessly slow, since `4bit` model downloading is not supported, and Unsloth's **inference is 2x faster**.

In [ ]:
if False:
    # I highly do NOT suggest - use Unsloth if possible
    from peft import AutoPeftModelForCausalLM
    from transformers import AutoTokenizer
    model = AutoPeftModelForCausalLM.from_pretrained(
        "lora_model", # YOUR MODEL YOU USED FOR TRAINING
        load_in_4bit = load_in_4bit,
    )
    tokenizer = AutoTokenizer.from_pretrained("lora_model")

### Saving to float16 for VLLM

We also support saving to `float16` directly. Select `merged_16bit` for float16 or `merged_4bit` for int4. We also allow `lora` adapters as a fallback. Use `push_to_hub_merged` to upload to your Hugging Face account! You can go to https://huggingface.co/settings/tokens for your personal tokens.

In [ ]:
# Merge to 16bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_16bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_16bit", token = "")

# Merge to 4bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_4bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_4bit", token = "")

# Just LoRA adapters
if False: model.save_pretrained_merged("model", tokenizer, save_method = "lora",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "lora", token = "")

Unsloth: You have 1 CPUs. Using `safe_serialization` is 10x slower.
We shall switch to Pytorch saving, which might take 3 minutes and not 30 minutes.
To force `safe_serialization`, set it to `None` instead.
Unsloth: Kaggle/Colab has limited disk space. We need to delete the downloaded
model which will save 4-16GB of disk space, allowing you to save on Kaggle/Colab.
Unsloth: Will remove a cached repo with size 15.1G


Unsloth: Merging 4bit and LoRA weights to 16bit...
Unsloth: Will use up to 3.99 out of 12.67 RAM for saving.
Unsloth: Saving model... This might take 5 minutes ...


100%|██████████| 28/28 [00:01<00:00, 27.83it/s]


Unsloth: Saving tokenizer... Done.
Unsloth: Saving model/pytorch_model-00001-of-00002.bin...
Unsloth: Saving model/pytorch_model-00002-of-00002.bin...
Done.


### GGUF / llama.cpp Conversion
To save to `GGUF` / `llama.cpp`, we support it natively now! We clone `llama.cpp` and we default save it to `q8_0`. We allow all methods like `q4_k_m`. Use `save_pretrained_gguf` for local saving and `push_to_hub_gguf` for uploading to HF.

Some supported quant methods (full list on our [Wiki page](https://github.com/unslothai/unsloth/wiki#gguf-quantization-options)):
* `q8_0` - Fast conversion. High resource use, but generally acceptable.
* `q4_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q4_K.
* `q5_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q5_K.

[**NEW**] To finetune and auto export to Ollama, try our [Ollama notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)

In [ ]:
# Save to 8bit Q8_0
if False: model.save_pretrained_gguf("model", tokenizer,)
# Remember to go to https://huggingface.co/settings/tokens for a token!
# And change hf to your username!
if False: model.push_to_hub_gguf("hf/model", tokenizer, token = "")

# Save to 16bit GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "f16")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "f16", token = "")

# Save to q4_k_m GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "q4_k_m")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "q4_k_m", token = "")

# Save to multiple GGUF options - much faster if you want multiple!
if False:
    model.push_to_hub_gguf(
        "hf/model", # Change hf to your username!
        tokenizer,
        quantization_method = ["q4_k_m", "q8_0", "q5_k_m",],
        token = "",
    )

Now, use the `model-unsloth.gguf` file or `model-unsloth-Q4_K_M.gguf` file in llama.cpp or a UI based system like Jan or Open WebUI. You can install Jan [here](https://github.com/janhq/jan) and Open WebUI [here](https://github.com/open-webui/open-webui)

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Train your own reasoning model - Llama GRPO notebook [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-GRPO.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️
</div>
